File: EngPractice/Service/ReadingService.cs
using EngPractice.Domain;
using Newtonsoft.Json;
using System.Text;

namespace EngPractice.Service
{
    public class ReadingService
    {
        private readonly HttpClient _httpClient;

        public ReadingService(HttpClient httpClient)
        {
            _httpClient = httpClient;
        }

        public async Task<ReadingPassageResponse> GenerateReadingPassage(EnglishLevel englishLevel, string apiKey)
        {
            if (!await HealthcheckService.Healthcheck(apiKey))
            {
                throw new Exception("API Key không hợp lệ.");
            }

            if (string.IsNullOrEmpty(apiKey))
            {
                throw new Exception("API Key không được để trống.");
            }

            var systemInstruction = Instructions.GetReadingPassageInstruction(englishLevel);

            var requestBody = new
            {
                contents = new[]
                {
                    new
                    {
                        role = "user",
                        parts = new[] { new { text = "Generate a reading passage for the specified English level." } }
                    }
                },
                system_instruction = new { parts = new[] { new { text = systemInstruction } } },
                generationConfig = new
                {
                    maxOutputTokens = 500,
                    temperature = 0.7
                }
            };

            var requestContent = new StringContent(
                JsonConvert.SerializeObject(requestBody),
                Encoding.UTF8,
                "application/json");

            var response = await _httpClient.PostAsync(
                $"https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key={apiKey}",
                requestContent);

            response.EnsureSuccessStatusCode();

            var responseBody = await response.Content.ReadAsStringAsync();
            Console.WriteLine($"Generate passage response: {responseBody}");

            var geminiResponse = JsonConvert.DeserializeObject<dynamic>(responseBody);
            string message = geminiResponse.candidates[0].content.parts[0].text;

            Console.WriteLine($"Message before processing: {message}");

            ReadingPassageResponse passageResponse;
            try
            {
                passageResponse = JsonConvert.DeserializeObject<ReadingPassageResponse>(message);
            }
            catch (JsonException jsonEx)
            {
                Console.WriteLine($"JSON deserialization error: {jsonEx.Message}");
                throw new Exception($"Lỗi khi xử lý phản hồi từ Gemini: {message}");
            }

            return passageResponse;
        }

        public async Task<EvaluationResponse> EvaluateGuess(string userGuess, string correctPhrase, string apiKey)
        {
            if (!await HealthcheckService.Healthcheck(apiKey))
            {
                throw new Exception("API Key không hợp lệ.");
            }

            if (string.IsNullOrEmpty(apiKey))
            {
                throw new Exception("API Key không được để trống.");
            }

            var systemInstruction = Instructions.GetEvaluationInstruction();

            var requestBody = new
            {
                contents = new[]
                {
                    new
                    {
                        role = "user",
                        parts = new[] { new { text = JsonConvert.SerializeObject(new { UserGuess = userGuess, CorrectPhrase = correctPhrase }) } }
                    }
                },
                system_instruction = new { parts = new[] { new { text = systemInstruction } } },
                generationConfig = new
                {
                    maxOutputTokens = 500,
                    temperature = 0.5
                }
            };

            var requestContent = new StringContent(
                JsonConvert.SerializeObject(requestBody),
                Encoding.UTF8,
                "application/json");

            var response = await _httpClient.PostAsync(
                $"https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key={apiKey}",
                requestContent);

            response.EnsureSuccessStatusCode();

            var responseBody = await response.Content.ReadAsStringAsync();
            Console.WriteLine($"Evaluate guess response: {responseBody}");

            var geminiResponse = JsonConvert.DeserializeObject<dynamic>(responseBody);
            string message = geminiResponse.candidates[0].content.parts[0].text;

            Console.WriteLine($"Message before processing: {message}");

            EvaluationResponse evaluationResponse;
            try
            {
                evaluationResponse = JsonConvert.DeserializeObject<EvaluationResponse>(message);
            }
            catch (JsonException jsonEx)
            {
                Console.WriteLine($"JSON deserialization error: {jsonEx.Message}");
                throw new Exception($"Lỗi khi xử lý phản hồi từ Gemini: {message}");
            }

            return evaluationResponse;
        }
    }

    public class ReadingPassageResponse
    {
        public string Description { get; set; }
        public string Phrase { get; set; }
    }

    public class EvaluationResponse
    {
        public int Accuracy { get; set; }
        public string Explanation { get; set; }
    }
}

File: backend-nestjs/src/reading/reading.service.ts
import { Injectable } from '@nestjs/common';
import { GoogleGenAI } from '@google/genai';
import { ReadingPrompts } from './reading.constants';
import { GenerateReadingDto } from './dto/generate-reading.dto';
import { EvaluateGuessDto } from './dto/evaluate-guess.dto';

@Injectable()
export class ReadingService {
  async generatePassage(dto: GenerateReadingDto) {
    try {
      const ai = new GoogleGenAI({ apiKey: dto.geminiApiKey });
      const prompt = ReadingPrompts.getReadingWord(
        dto.englishLevel,
        dto.usedDescriptions,
      );

      const readingSchema = {
        type: 'object',
        properties: {
          description: { type: 'string', description: 'Câu đố bằng tiếng Anh' },
          translation: { type: 'string', description: 'Dịch nghĩa câu đố' },
          phrase: { type: 'string', description: 'Từ khóa đáp án (1 từ)' },
          phraseTranslation: {
            type: 'string',
            description: 'Nghĩa tiếng Việt của đáp án',
          },
        },
        required: ['description', 'translation', 'phrase', 'phraseTranslation'],
      };

      const response = await ai.models.generateContent({
        model: 'gemini-2.5-flash-lite',
        contents: [{ role: 'user', parts: [{ text: prompt }] }],
        config: {
          responseMimeType: 'application/json',
          responseSchema: readingSchema,
          temperature: 0.7,
        },
      });

      const text =
        response.text || 'Có lỗi xảy ra ở phía máy chủ, vui lòng thử lại sau.';
      const parsed = JSON.parse(text);

      return {
        description: parsed.description,
        translation: parsed.translation,
        phrase: parsed.phrase,
        phraseTranslation: parsed.phraseTranslation,
      };
    } catch (e) {
      console.error('Generate Reading Error:', e);
      throw new Error('Không thể tạo câu hỏi lúc này.');
    }
  }

  async evaluateGuess(dto: EvaluateGuessDto) {
    try {
      const ai = new GoogleGenAI({ apiKey: dto.geminiApiKey });
      const prompt = ReadingPrompts.getEvaluation(
        dto.userGuess,
        dto.correctPhrase,
      );

      const evaluationSchema = {
        type: 'object',
        properties: {
          accuracy: {
            type: 'integer',
            description: 'Điểm số từ 0 đến 100 dựa trên độ chính xác.',
          },
          explanation: {
            type: 'string',
            description: 'Giải thích ngắn gọn bằng tiếng Việt.',
          },
          explanationEnglish: {
            type: 'string',
            description: 'Giải thích ngắn gọn bằng tiếng Anh.',
          },
        },
        required: ['accuracy', 'explanation', 'explanationEnglish'],
      };

      const response = await ai.models.generateContent({
        model: 'gemini-2.5-flash',
        contents: [{ role: 'user', parts: [{ text: prompt }] }],
        config: {
          responseMimeType: 'application/json',
          responseSchema: evaluationSchema,
          temperature: 0.3,
        },
      });

      const text =
        response.text || 'Có lỗi xảy ra ở phía máy chủ, vui lòng thử lại sau.';
      return JSON.parse(text);
    } catch (error) {
      console.error('Evaluation Error:', error);
      return {
        accuracy: 0,
        explanation: 'Không thể đánh giá vào lúc này. Vui lòng thử lại.',
        explanationEnglish: 'Cannot evaluate at this time. Please try again.',
      };
    }
  }
}
